c error when dump shared struct

v doctor:
```
OS: linux, Ubuntu 20.04.4 LTS
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i5-3320M CPU @ 2.60GHz
CC version: cc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0

getwd: /home/woody/sources/v/bugs
vmodules: /home/woody/.vmodules
vroot: /home/woody/sources/v/v
vexe: /home/woody/sources/v/v/v
vexe mtime: 2022-08-27 19:39:38
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.0 d08edf8.d75c62b

Git version: git version 2.25.1
Git vroot status: weekly.2022.34-38-gd75c62b8
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 827f7452
```

code, dump_atomic.v :
```
module main

struct AtomicStruct{
	a int
}

fn main() {
	shared inst := AtomicStruct{a: 1}
	lock inst {
		dump(inst)
	}
}

```

```
v dump_atomic.v 
==================
/tmp/v_1000/dump_atomic.18181863581555602909.tmp.c:3332: error: cannot convert 'struct main__AtomicStruct' to 'struct __shared__main__AtomicStruct'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang


```

--- a/vlib/v/gen/c/dumpexpr.v
+++ b/vlib/v/gen/c/dumpexpr.v
@@ -14,7 +14,13 @@ fn (mut g Gen) dump_expr(node ast.DumpExpr) {
 	}
 	dump_fn_name := '_v_dump_expr_$node.cname' + (if node.expr_type.is_ptr() { '_ptr' } else { '' })
 	g.write(' ${dump_fn_name}(${ctoslit(fpath)}, $line, $sexpr, ')
-	g.expr(node.expr)
+	if node.expr_type.has_flag(.shared_f) {
+		g.write('&')
+		g.expr(node.expr)
+		g.write('->val')
+	} else {
+		g.expr(node.expr)
+	}
 	g.write(' )')
 }
 
@@ -27,7 +33,7 @@ fn (mut g Gen) dump_expr_definitions() {
 		_, str_method_expects_ptr, _ := dump_sym.str_method_info()
 		is_ptr := ast.Type(dump_type).is_ptr()
 		deref, _ := deref_kind(str_method_expects_ptr, is_ptr, dump_type)
-		to_string_fn_name := g.get_str_fn(dump_type)
+		to_string_fn_name := g.get_str_fn(ast.Type(dump_type).clear_flag(.shared_f))
 		ptr_asterisk := if is_ptr { '*' } else { '' }
 		mut str_dumparg_type := '$cname$ptr_asterisk'
 		if dump_sym.kind == .function {

--- a/vlib/v/tests/inout/dump_shared_arg.out
+++ b/vlib/v/tests/inout/dump_shared_arg.out
@@ -1,0 +1,3 @@
+[vlib/v/tests/inout/dump_shared_arg.vv:12] inst: &AtomicStruct{
+    a: 1
+}

--- a/vlib/v/tests/inout/dump_shared_arg.vv
+++ b/vlib/v/tests/inout/dump_shared_arg.vv
@@ -1,0 +1,14 @@
+module main
+
+struct AtomicStruct {
+	a int
+}
+
+fn main() {
+	shared inst := AtomicStruct{
+		a: 1
+	}
+	lock inst {
+		dump(inst)
+	}
+}

